### PR TITLE
[PVR] Implement smart direct channel number input

### PR DESCRIPTION
**Pull request by ksooo**
Feature requested in the forum: https://forum.kodi.tv/showthread.php?tid=331022

I just exteded the feature requested there to be even more clever. ;-)

If direct channel number input is used, now on every typed character a check will be done, whether the already typed in part of the channel number is sufficient to identify the target channel number. If so, the number will be completed automatically and the channel switch will be done immediately, without need to press OK/Enter. Otherwise, the input handler will wait for the next character.

Example: possible channels: 1 2 10 11 33

Type "3" => immediately auto switch to "33"
Type "1" => type nothing until handler timeout =>auto switch to "1" after timeout
Type "1" => hit OK/Enter => switch to "1"
Type "1" => type another "1" before handler timeout => immediately switch to "11"
Type "8" => type nothing until handler timeout => nothing happens (invalid channel number)

I runtime-tested the changes with kodi latest master on macOS.

@Jalle19 mind taking a look.

### CODE REVIEW

**Comment by Jalle19** [comment]
@ksooo the idea sounds nice but I think this fails if you configure Kodi to "use backend channel numbers". Then you can have just 40 channels but the last one could have number 2301, so this logic (at least the one described on the forum) would fail.

**Comment by ksooo** [comment]
> the idea sounds nice but I think this fails if you configure Kodi to "use backend channel numbers". Then you can have just 40 channels but the last one could have number 2301, so this logic (at least the one described on the forum) would fail.

It does not fail. I did not implement the (as you spotted not reliably working) algorithm suggested in the forum, but did it "right".

File: xbmc/pvr/PVRChannelNumberInputHandler.cpp [hunk 1]
```diff
@@ -80,9 +102,42 @@ void CPVRChannelNumberInputHandler::AppendChannelNumberCharacter(char cCharacter
...
     m_inputBuffer.erase(m_inputBuffer.begin());
+    m_label = m_inputBuffer;
+  }
+  else if (m_inputBuffer.empty())
+  {
+    m_sortedChannelNumbers.clear();
+    GetChannelNumbers(m_sortedChannelNumbers);
+
+    // sort channel numbers alphabetically
+    std::sort(m_sortedChannelNumbers.begin(), m_sortedChannelNumbers.end());
```
Comment (Jalle19) [inline xbmc/pvr/PVRChannelNumberInputHandler.cpp]: I have a feeling this fails since the numbers are strings, we would have to do a "natural sort"

Comment (Jalle19) [reply xbmc/pvr/PVRChannelNumberInputHandler.cpp]: Also what about sub channel numbers?

Comment (ksooo) [reply xbmc/pvr/PVRChannelNumberInputHandler.cpp]: No, it works perfect. I tested it even with subchannel numbers.

Comment (ksooo) [reply xbmc/pvr/PVRChannelNumberInputHandler.cpp]: screen shot 2018-04-25 at 20 45 42

This is how it looks after std:.sort

**Comment by ksooo** [comment]
Jenkins error is unrelated.

**Review by Jalle19 — APPROVED** [review]
If it works, +1

**Pull request merged**